odoo/addons/base/models/res_users.py

```diff
@@ -2336,48 +2341,124 @@ def _check_credentials(self, *, scope, key):
+        [name, self.env.user.id, scope, expiration_date, KEY_CRYPT_CONTEXT.hash(k), k[:INDEX_SIZE]])
 
         ip = request.httprequest.environ['REMOTE_ADDR'] if request else 'n/a'
         _logger.info("%s generated: scope: <%s> for '%s' (#%s) from %s",
             self._description, scope, self.env.user.login, self.env.uid, ip)
 
         return k
 
+    @api.autovacuum
+    def _gc_user_apikeys(self):
+        self.env.cr.execute(SQL("""
+            DELETE FROM %s
+            WHERE
+                expiration_date IS NOT NULL AND
+                expiration_date < now() at time zone 'utc'
+        """, SQL.identifier(self._table)))
+        _logger.info("GC %r delete %d entries", self._name, self.env.cr.rowcount)
+
 class APIKeyDescription(models.TransientModel):
     _name = 'res.users.apikeys.description'
     _description = 'API Key Description'
 
+    def _selection_duration(self):
+        # duration value is a string representing the number of days.
+        durations = [
+            ('1', '1 Day'),
+            ('7', '1 Week'),
+            ('30', '1 Month'),
+            ('90', '3 Months'),
+            ('180', '6 Months'),
+            ('365', '1 Year'),
+            ('-1', 'Persistent Key'),  # Magic value to detect an infinite duration
+            ('0', 'Custom Date'),  # Will force the user to enter a date manually
+        ]
+        if self.env.is_system():
+            return durations
+        max_duration = max(group.api_key_duration for group in self.env.user.groups_id) or 1.0
+        return list(filter(lambda duration: 0 <= int(duration[0]) <= max_duration, durations))
+
     name = fields.Char("Description", required=True)
+    duration = fields.Selection(selection='_selection_duration', string='Duration', required=True,
+        default=lambda self: self._selection_duration()[0][0])
+    custom_expiration_date = fields.Date('Expiration Date', default=fields.Date.today() + datetime.timedelta(days=1))
+
+    @api.onchange('custom_expiration_date')
+    def _onchange_custom_expiration_date(self):
+        duration = (self.custom_expiration_date - fields.Date.today()).days
+        try:
+            self.env['res.users.apikeys']._check_duration(duration)
+        except UserError as error:
+            warning = {
+                'type': 'notification',
+                'title': _('The API key duration is not correct.'),
+                'message': error.args[0]
+            }
+            return {'warning': warning}
 
     @check_identity
     def make_key(self):
         # only create keys for users who can delete their keys
         self.check_access_make_key()
 
         description = self.sudo()
-        k = self.env['res.users.apikeys']._generate(None, self.sudo().name)
+        duration = int(description.duration) \
+            or (max(self.custom_expiration_date, fields.Date.today()) - fields.Date.today()).days \
+            or 1.0
```

Comment:
This seems overly complicated computation

And in addition you convert the expiration_date to a duration, and then `_generate` re-converts back the duration to an expiration date.

Wouldn't it be easier to pass the expiration date directly to `_generate` ?

It makes it "a bit" more complicated for people calling it, as they have to compute a date.
But
- It avoids the back and forth duration <-> date
- It makes the logical code clearer
- it offers more flexibility. For instance, there is no longer a need to convert the variable `TRUSTED_DEVICE_AGE` to days instead of second.

`custom_expiration_date` on the wizard can also be renamed simplier `expiration_date`.

```diff
diff --git a/addons/auth_totp/controllers/home.py b/addons/auth_totp/controllers/home.py
index ea8ecdc418f9..1cefc9a00421 100644
--- a/addons/auth_totp/controllers/home.py
+++ b/addons/auth_totp/controllers/home.py
@@ -1,13 +1,15 @@
 # -*- coding: utf-8 -*-
 import re
 
+from datetime import datetime
+
 from odoo import http, _
 from odoo.exceptions import AccessDenied
 from odoo.http import request
 from odoo.addons.web.controllers import home as web_home
 
 TRUSTED_DEVICE_COOKIE = 'td_id'
-TRUSTED_DEVICE_DURATION = 90  # 90 days expiration
+TRUSTED_DEVICE_AGE = 90*86400 # 90 days expiration
 
 
 class Home(web_home.Home):
@@ -58,11 +60,15 @@ class Home(web_home.Home):
                     if request.geoip.city.name:
                         name += f" ({request.geoip.city.name}, {request.geoip.country_name})"
 
-                    key = request.env['auth_totp.device'].sudo()._generate("browser", name, TRUSTED_DEVICE_DURATION)
+                    key = request.env['auth_totp.device'].sudo()._generate(
+                        "browser",
+                        name,
+                        datetime.now() + datetime.timedelta(seconds=TRUSTED_DEVICE_AGE)
+                    )
                     response.set_cookie(
                         key=TRUSTED_DEVICE_COOKIE,
                         value=key,
-                        max_age=TRUSTED_DEVICE_DURATION * 86400,
+                        max_age=TRUSTED_DEVICE_AGE,
                         httponly=True,
                         samesite='Lax'
                     )
diff --git a/addons/auth_totp_mail/models/auth_totp_device.py b/addons/auth_totp_mail/models/auth_totp_device.py
index bc79748f412b..f537426ec604 100644
--- a/addons/auth_totp_mail/models/auth_totp_device.py
+++ b/addons/auth_totp_mail/models/auth_totp_device.py
@@ -22,12 +22,12 @@ class AuthTotpDevice(models.Model):
 
         return super().unlink()
 
-    def _generate(self, scope, name, duration):
+    def _generate(self, scope, name, expiration_date):
         """ Notify users when trusted devices are added onto their account.
         We override this method instead of 'create' as those records are inserted directly into the
         database using raw SQL. """
 
-        res = super()._generate(scope, name, duration)
+        res = super()._generate(scope, name, expiration_date)
 
         self.env.user._notify_security_setting_update(
             _("Security Update: Device Added"),
diff --git a/addons/auth_totp_mail/tests/test_notify_security_update_totp.py b/addons/auth_totp_mail/tests/test_notify_security_update_totp.py
index 260bbd45fc65..96e857046a25 100644
--- a/addons/auth_totp_mail/tests/test_notify_security_update_totp.py
+++ b/addons/auth_totp_mail/tests/test_notify_security_update_totp.py
@@ -1,7 +1,9 @@
 # -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
-from odoo.addons.auth_totp.controllers.home import TRUSTED_DEVICE_DURATION
+from datetime import datetime
+
+from odoo.addons.auth_totp.controllers.home import TRUSTED_DEVICE_AGE
 from odoo.addons.mail.tests.test_res_users import TestNotifySecurityUpdate
 from odoo.tests import users
 
@@ -29,7 +31,11 @@ class TestNotifySecurityUpdateTotp(TestNotifySecurityUpdate):
         """ Make sure we notify the user when TOTP trusted devices are added/removed on his account. """
         recipients = [self.env.user.email_formatted]
         with self.mock_mail_gateway():
-            self.env['auth_totp.device'].sudo()._generate('trusted_device_chrome', 'Chrome on Windows', TRUSTED_DEVICE_DURATION)
+            self.env['auth_totp.device'].sudo()._generate(
+                'trusted_device_chrome',
+                'Chrome on Windows',
+                datetime.now() + datetime.timedelta(seconds=TRUSTED_DEVICE_AGE),
+            )
 
         self.assertMailMailWEmails(recipients, 'outgoing', fields_values={
             'subject': 'Security Update: Device Added',
@@ -37,7 +43,11 @@ class TestNotifySecurityUpdateTotp(TestNotifySecurityUpdate):
 
         # generating a key outside of the 'auth_totp.device' model should however not notify
         with self.mock_mail_gateway():
-            self.env['res.users.apikeys']._generate('new_api_key', 'New Key', 1)
+            self.env['res.users.apikeys']._generate(
+                'new_api_key',
+                'New Key',
+                datetime.now() + datetime.timedelta(days=1),
+            )
         self.assertNotSentEmail(recipients)
 
         # now remove the key using the user's relationship
diff --git a/addons/mail_plugin/controllers/authenticate.py b/addons/mail_plugin/controllers/authenticate.py
index 81e865861841..9f93b29e54e0 100644
--- a/addons/mail_plugin/controllers/authenticate.py
+++ b/addons/mail_plugin/controllers/authenticate.py
@@ -73,7 +73,11 @@ class Authenticate(http.Controller):
             return {"error": "Invalid code"}
         request.update_env(user=auth_message['uid'])
         scope = 'odoo.plugin.' + auth_message.get('scope', '')
-        api_key = request.env['res.users.apikeys']._generate(scope, auth_message['name'], 1)
+        api_key = request.env['res.users.apikeys']._generate(
+            scope,
+            auth_message['name'],
+            datetime.now() + datetime.timedelta(days=1),
+        )
         return {'access_token': api_key}
 
     def _get_auth_code_data(self, auth_code):
diff --git a/addons/portal/tests/test_portal.py b/addons/portal/tests/test_portal.py
index 02ea7ddc0b6c..004eb7066a5b 100644
--- a/addons/portal/tests/test_portal.py
+++ b/addons/portal/tests/test_portal.py
@@ -1,3 +1,5 @@
+from datetime import datetime
+
 from odoo import Command
 from odoo.http import Request
 from odoo.tests.common import HttpCase, tagged
@@ -15,7 +17,11 @@ class TestUsersHttp(HttpCase):
             'password': login,
             'groups_id': [Command.set([self.env.ref('base.group_portal').id])],
         })
-        self.env['res.users.apikeys'].with_user(portal_user)._generate(None, 'Portal API Key', 1)
+        self.env['res.users.apikeys'].with_user(portal_user)._generate(
+            None,
+            'Portal API Key',
+            datetime.now() + datetime.timedelta(days=1),
+        )
         self.assertTrue(portal_user.api_key_ids)
 
         # Request the deactivation of the portal account as portal through the route meant for this purpose
diff --git a/odoo/addons/base/models/res_users.py b/odoo/addons/base/models/res_users.py
index e19f7acf952c..39872db78436 100644
--- a/odoo/addons/base/models/res_users.py
+++ b/odoo/addons/base/models/res_users.py
@@ -2354,21 +2354,19 @@ class APIKeys(models.Model):
             if KEY_CRYPT_CONTEXT.verify(key, current_key):
                 return user_id
 
-    def _check_duration(self, duration):
+    def _check_expiration_date(self, date):
         # To be in a sudoed environment or to be an administrator
-        # to create a persistent key (duration is -1) or
+        # to create a persistent key (no expiration date) or
         # to exceed the maximum duration determined by the user's privileges.
-        if duration == -1 and self.env.is_system():
-            return
-        if duration <= 0:
-            raise UserError(_("Duration must be a strictly positive number."))
         if self.env.is_system():
             return
+        if not date:
+            raise UserError(_("The API key must have an expiration date"))
         max_duration = max(group.api_key_duration for group in self.env.user.groups_id) or 1.0
-        if duration > max_duration:
+        if date > datetime.now() + datetime.timedelta(days=max_duration):
             raise UserError(_("You cannot exceed %(duration)s days.", duration=max_duration))
 
-    def _generate(self, scope, name, duration):
+    def _generate(self, scope, name, expiration_date):
         """Generates an api key.
         :param str scope: the scope of the key. If None, the key will give access to any rpc.
         :param str name: the name of the key, mainly intended to be displayed in the UI.
@@ -2381,10 +2379,7 @@ class APIKeys(models.Model):
         For a persistent key (infinite duration),
         the value must be '-1' (sudo or administrator).
         """
-        self._check_duration(duration)
-        expiration_date = None
-        if duration > 0:
-            expiration_date = datetime.datetime.now() + datetime.timedelta(days=duration)
+        self._check_expiration_date(expiration_date)
         # no need to clear the LRU when *adding* a key, only when removing
         k = binascii.hexlify(os.urandom(API_KEY_SIZE)).decode()
         self.env.cr.execute("""
@@ -2392,7 +2387,7 @@ class APIKeys(models.Model):
         VALUES (%s, %s, %s, %s, %s, %s)
         RETURNING id
         """.format(table=self._table),
-        [name, self.env.user.id, scope, expiration_date, KEY_CRYPT_CONTEXT.hash(k), k[:INDEX_SIZE]])
+        [name, self.env.user.id, scope, expiration_date or None, KEY_CRYPT_CONTEXT.hash(k), k[:INDEX_SIZE]])
 
         ip = request.httprequest.environ['REMOTE_ADDR'] if request else 'n/a'
         _logger.info("%s generated: scope: <%s> for '%s' (#%s) from %s",
@@ -2423,8 +2418,8 @@ class APIKeyDescription(models.TransientModel):
             ('90', '3 Months'),
             ('180', '6 Months'),
             ('365', '1 Year'),
-            ('-1', 'Persistent Key'),  # Magic value to detect an infinite duration
-            ('0', 'Custom Date'),  # Will force the user to enter a date manually
+            ('0', 'Persistent Key'),  # Magic value to detect an infinite duration
+            ('-1', 'Custom Date'),  # Will force the user to enter a date manually
         ]
         if self.env.is_system():
             return durations
@@ -2432,15 +2427,27 @@ class APIKeyDescription(models.TransientModel):
         return list(filter(lambda duration: 0 <= int(duration[0]) <= max_duration, durations))
 
     name = fields.Char("Description", required=True)
-    duration = fields.Selection(selection='_selection_duration', string='Duration', required=True,
-        default=lambda self: self._selection_duration()[0][0])
-    custom_expiration_date = fields.Date('Expiration Date', default=fields.Date.today() + datetime.timedelta(days=1))
+    duration = fields.Selection(
+        selection='_selection_duration', string='Duration', required=True,
+        default=lambda self: self._selection_duration()[0][0],
+    )
+    expiration_date = fields.Datetime('Expiration Date', compute='_compute_expiration_date', store=True, readonly=False)
+
+    @api.depends('duration')
+    def _compute_expiration_date(self):
+        for record in self:
+            duration = int(record.duration)
+            if duration >= 0:
+                record.expiration_date = (
+                    fields.Date.today() + datetime.timedelta(days=duration)
+                    if int(record.duration)
+                    else None
+                )
 
-    @api.onchange('custom_expiration_date')
-    def _onchange_custom_expiration_date(self):
-        duration = (self.custom_expiration_date - fields.Date.today()).days
+    @api.onchange('expiration_date')
+    def _onchange_expiration_date(self):
         try:
-            self.env['res.users.apikeys']._check_duration(duration)
+            self.env['res.users.apikeys']._check_expiration_date(self.expiration_date)
         except UserError as error:
             warning = {
                 'type': 'notification',
@@ -2455,10 +2462,7 @@ class APIKeyDescription(models.TransientModel):
         self.check_access_make_key()
 
         description = self.sudo()
-        duration = int(description.duration) \
-            or (max(self.custom_expiration_date, fields.Date.today()) - fields.Date.today()).days \
-            or 1.0
-        k = self.env['res.users.apikeys']._generate(None, description.name, duration)
+        k = self.env['res.users.apikeys']._generate(None, description.name, self.expiration_date)
         description.unlink()
 
         return {
diff --git a/odoo/addons/base/views/res_users_views.xml b/odoo/addons/base/views/res_users_views.xml
index a6ed009ecd7e..5495c350261a 100644
--- a/odoo/addons/base/views/res_users_views.xml
+++ b/odoo/addons/base/views/res_users_views.xml
@@ -376,7 +376,7 @@
                         Give a duration for the key's validity
                     </h3>
                     <field name="duration"/>
-                    <field name="custom_expiration_date" invisible="duration != '0'"/>
+                    <field name="expiration_date" invisible="duration != '-1'"/>
                     <p>
                         The key will be deleted once this period has elapsed.
                     </p>

```